dashboard/app/helpers/levels_helper.rb

```diff
@@ -487,6 +488,31 @@ def azure_speech_service_options
       response = http.request(request)
       speech_service_options[:azureSpeechServiceToken] = response.body
       speech_service_options[:azureSpeechServiceRegion] = CDO.azure_speech_service_region
+      uri = URI.parse("https://westus.tts.speech.microsoft.com/cognitiveservices/voices/list")
+      header = {'Authorization': 'Bearer ' + response.body}
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl = true
+      http.verify_mode = OpenSSL::SSL::VERIFY_PEER
+      request = Net::HTTP::Get.new(uri.request_uri, header)
+      response = http.request(request)
+      all_voices = JSON.parse(response.body)
+      language_dictionary = {}
+      all_voices.each do |voice|
+        if language_dictionary[voice["Locale"]].nil?
+          language_dictionary[voice["Locale"]] = {}
+          language_dictionary[voice["Locale"]][voice["Gender"].downcase] = voice["ShortName"]
+        elsif language_dictionary[voice["Locale"]][voice["Gender"].downcase].nil?
+          language_dictionary[voice["Locale"]][voice["Gender"].downcase] = voice["ShortName"]
+        end
+      end
+      all_languages = Languages.get_locale.map {|language| language[:locale_s]}
+      language_dictionary.keys.each do |language|
+        if language_dictionary[language].size < 2 || !(all_languages.include? language)
+          language_dictionary.delete(language)
+        end
+      end
+      language_dictionary = language_dictionary.transform_keys {|locale| Languages.get_native_name_by_locale(locale)}
```

Comment:
i think we could refactor this so that we only have to iterate over `all_voices` once in order to build our `language_dictionary` by doing a few checks at once:

1. do we have a native name for this locale?
2. do we have male and female voices for this locale?
3. if 1 & 2 are true, then add the male/female voices to our `language_dictionary` with the key already transformed to its native name

Reply:
I can check for the native name earlier. However, to find out if we have both male and female voices for the locale, we have to iterate through all_voices (it's an array of json objects and some languages have more than one voice for each gender so there's no easy rule to tell if both options are available).

Reply:
ah, yeah, having multiples of the same gender for a locale definitely makes that more difficult. there are still some ruby secrets we can use to simplify this code though:

```ruby
language_dictionary = {}
all_voices.each do |voice|
  # NOTE: The next line will break if we don't support the requested locale
  native_locale_name = Languages.get_native_name_by_locale(voice["Locale"])
  next unless native_locale_name
  language_dictionary[native_locale_name] ||= {}
  language_dictionary[native_locale_name][voice["Gender"].downcase] ||= voice["ShortName"]
end

language_dictionary.delete_if {|language, voices| voices.length < 2}

speech_service_options[:azureSpeechServiceLanguages] = language_dictionary
```

re: the `NOTE` comment above -- the `Languages.get_native_name_by_locale` will need to be refactored for this to work (it currently breaks if you give it a locale that isn't supported), but i think that's a necessary refactor to do anyways.